### ISSUE TITLE
Why filter by missing rate iteratively?

### ISSUE THREAD

**Issue by sennpuuki** [JoniColeman/gwas_scripts#5]
Thanks for your paper and the code. I'm learning to QC the chip data and I am confused by the filtering steps in script: Iterative_Missingness.sh
You iteratively remove the SNPs and samples by a sequential threshold (0.1, 0.09, 0.08, ..., 0.01 when the arguments are set as 90 99 1) with this script while my previous understanding of filtering by missing rate is to remove SNPs and samples by one rate (say 0.05 or 0.1). What is the point of doing the consequent filtering and how is that better than just remove by a single threshold? Wouldn't this be too stringent in general?
Looking forward to your response. Thanks in advance!

**Comment by JoniColeman** [comment]
I'm glad you are finding the code and paper helpful.

**TL;DR: iterative missingness retains individuals at the expense of an increased loss of variants. I think this is preferable to losing individuals.**

There are three basic approaches to removing data with low call-rates: 

1. PLINK's native behaviour if you pass both --mind and --geno is to remove low call-rate individuals and then to remove low call-rate variants (see https://www.cog-genomics.org/plink/1.9/order). 
2. You could reverse this approach by applying --geno in one run of PLINK, and then apply --mind in the next 
3. You could apply iterative missingness, removing the lowest call-rate variants then individuals, then the next lowest etc.  

Imagine a dataset like this one, where "." is a missing variant and "1" is any genotype call:

> IID | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> Ali | . | . | 1 | 1 | 1 | 1 | 1 | 1 | 1 | . | 7/10
> Beth | 1 | . | 1 | 1 | 1 | 1 | . | 1 | 1 | 1 | 8/10
> Carol | 1 | 1 | . | 1 | 1 | 1 | . | . | 1 | 1 | 7/10
> Dave | 1 | . | 1 | 1 | . | . | . | 1 | . | 1 | 5/10
> Ed | 1 | 1 | 1 | 1 | 1 | 1 | . | 1 | 1 | 1 | 9/10
> Total | 4/5 | 2/5 | 4/5 | 5/5 | 4/5 | 4/5 | 1/5 | 4/5 |  4/5 | 4/5 |

Hopefully it's apparent from a cursory inspection that:

- Dave's data is unreliable and we don't want to keep it. 
- Variants 2 and 7 have been repeatedly missed, and these variants should also be excluded.

Ideally, we'd do this kind of inspection for all data, but that's not practical once you get to a dataset of any useful sample size or genome coverage, hence the heuristic of dropping individuals and variants for missingness. 

Imagine applying each method to the data such that you are aiming for 80% call-rate in individuals and variants.
 
1. (Mind 0.01 then geno 0.01)

- Everyone except Ed and Beth has individual level call-rate <80%. So they all get dropped.

> IID | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> ~~Ali~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~7/10~~
> Beth | 1 | . | 1 | 1 | 1 | 1 | . | 1 | 1 | 1 | 8/10
> ~~Carol~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~7/10~~
> ~~Dave~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~5/10~~
> Ed | 1 | 1 | 1 | 1 | 1 | 1 | . | 1 | 1 | 1 | 9/10
> Total | 2/2 | 1/2 | 2/2 | 2/2 | 2/2| 2/2 | 0/2 | 2/2 |  2/2 | 2/2 |


- Variants 2 and 7 then have < 80% call rate, and all other variants have 100% call rate.

> IID | 1 | ~~2~~ | 3 | 4 | 5 | 6 | ~~7~~ | 8 | 9 | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> ~~Ali~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~7/10~~
> Beth | 1 | ~~.~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> ~~Carol~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~7/10~~
> ~~Dave~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~5/10~~
> Ed | 1 | ~~1~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Total | 2/2 | ~~1/2~~ | 2/2 | 2/2 | 2/2| 2/2 | ~~0/2~~ | 2/2 |  2/2 | 2/2 |

This is not a good result - we've lost more individuals than was necessary from inspection.

2. (Geno 0.01 then mind 0.01)

- Variants 2 and 7 have call-rate < 80%, so they are dropped.

> IID | 1 | ~~2~~ | 3 | 4 | 5 | 6 | ~~7~~ | 8 | 9 | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> Ali | . | ~~.~~ | 1 | 1 | 1 | 1 | ~~1~~ | 1 | 1 | . | 6/8
> Beth | 1 | ~~.~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Carol | 1 | ~~1~~ | . | 1 | 1 | 1 | ~~.~~ | . | 1 | 1 | 6/8
> Dave | 1 | ~~.~~ | 1 | 1 | . | . | ~~.~~ | 1 | . | 1 | 5/8
> Ed | 1 | ~~1~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Total | 4/5 | ~~2/5~~ | 4/5 | 5/5 | 4/5 | 4/5 | ~~1/5~~ | 4/5 | 4/5 | 4/5 |

- Ali, Carol and Dave now have call rates < 80%, so drop them.

> IID | 1 | ~~2~~ | 3 | 4 | 5 | 6 | ~~7~~ | 8 | ~~9~~ | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> ~~Ali~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~6/8~~
> Beth | 1 | ~~.~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> ~~Carol~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~6/8~~
> ~~Dave~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~5/8~~
> Ed | 1 | ~~1~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Total | 2/2 | ~~2/5~~ | 2/2 | 2/2 | 2/2 | 2/2 | ~~1/5~~ | 2/2 | 2/2 | 2/2 |

Same result as 1.

3. (Worst geno, then worst mind, iterative missingness)

- The lowest call-rate across individuals and variants is 20% (Variant 7), so start from there and move up in sets of 10% (I have omitted iterations where nothing changes below).

- Remove variants with call rate 20% or lower (Variant 7)

> IID | 1 | 2 | 3 | 4 | 5 | 6 | ~~7~~ | 8 | 9 | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> Ali | . | . | 1 | 1 | 1 | 1 | ~~1~~ | 1 | 1 | . | 6/9
> Beth | 1 | . | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/9
> Carol | 1 | 1 | . | 1 | 1 | 1 | ~~.~~ | . | 1 | 1 | 7/9
> Dave | 1 | . | 1 | 1 | . | . | ~~.~~ | 1 | . | 1 | 5/9
> Ed | 1 | 1 | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 9/9
> Total | 4/5 | 2/5 | 4/5 | 5/5 | 4/5 | 4/5 | ~~1/5~~ | 4/5 |  4/5 | 4/5 |

- Remove variants with call rate 40% or lower (Variant 2)

> IID | 1 | ~~2~~ | 3 | 4 | 5 | 6 | ~~7~~ | 8 | 9 | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> Ali | . | ~~.~~ | 1 | 1 | 1 | 1 | ~~1~~ | 1 | 1 | . | 6/8
> Beth | 1 | ~~.~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Carol | 1 | ~~1~~ | . | 1 | 1 | 1 | ~~.~~ | . | 1 | 1 | 6/8
> Dave | 1 | ~~.~~ | 1 | 1 | . | . | ~~.~~ | 1 | . | 1 | 5/8
> Ed | 1 | ~~1~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Total | 4/5 | ~~2/5~~ | 4/5 | 5/5 | 4/5 | 4/5 | ~~1/5~~ | 4/5 |  4/5 | 4/5 |

- Remove individuals with call rate 70% or lower (Dave)

> IID | 1 | ~~2~~ | 3 | 4 | 5 | 6 | ~~7~~ | 8 | 9 | 10 | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> Ali | . | ~~.~~ | 1 | 1 | 1 | 1 | ~~1~~ | 1 | 1 | . | 6/8
> Beth | 1 | ~~.~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Carol | 1 | ~~1~~ | . | 1 | 1 | 1 | ~~.~~ | . | 1 | 1 | 6/8
> ~~Dave~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~5/8~~
> Ed | 1 | ~~1~~ | 1 | 1 | 1 | 1 | ~~.~~ | 1 | 1 | 1 | 8/8
> Total | 3/4 | ~~2/5~~ | 3/4 | 4/4 | 4/4 | 4/4 | ~~1/5~~ | 3/4 |  4/4 | 3/4 |

- Remove variants with call rate lower than 80% (Variants 1, 3, 8, 10)

> IID | ~~1~~ | ~~2~~ | ~~3~~ | 4 | 5 | 6 | ~~7~~ | ~~8~~ | 9 | ~~10~~ | Total
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
> Ali | ~~.~~ | ~~.~~ | ~~1~~ | 1 | 1 | 1 | ~~1~~ | ~~1~~ | 1 | ~~.~~ | 4/4
> Beth | ~~1~~ | ~~.~~ | ~~1~~ | 1 | 1 | 1 | ~~.~~ | ~~1~~ | 1 | ~~1~~ | 4/4
> Carol | ~~1~~ | ~~1~~ | ~~.~~ | 1 | 1 | 1 | ~~.~~ | ~~.~~ | 1 | ~~1~~ | 4/4
> ~~Dave~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~1~~ | ~~.~~ | ~~.~~ | ~~.~~ | ~~1~~ | ~~.~~ | ~~1~~ | ~~5/8~~
> Ed | ~~1~~ | ~~1~~ | ~~1~~ | 1 | 1 | 1 | ~~.~~ | ~~1~~ | 1 | ~~1~~ | 4/4
> Total | ~~3/4~~ | ~~2/5~~ | ~~3/4~~ | 4/4 | 4/4 | 4/4 | ~~1/5~~ | ~~3/4~~ |  4/4 | ~~3/4~~ |

We now have complete data across four variants for everyone except Dave - compared to the methods above, we've retained two extra individuals for the cost of four variants.

This is obviously a toy example I've rigged to show the point, but applications in real data show the same thing - iterative missingness retains individuals at the expense of variants. 

Is it a concern to lose extra variants? I don't personally think it is, because the majority of applications are going to then impute the resulting genotype data, which will end up replacing the lost genotypes in many cases. But this is, of course, only one approach - I think it is useful, but it is way more important to understand the necessity of removing bad variants and individuals. If iterative missingness is not appropriate for your data or for the application you have in mind, then don't use it.
